### ISSUE TITLE
Tap to pay on NFC mobile cause error: Stripe Terminal ConnectionToken has expired. Please download a new one using v1/terminal/connection_tokens

### ISSUE THREAD

**Issue by ajithv11** [stripe/stripe-terminal-android#344]
While running the tap to pay on my mobile device with NFC support, at some point I got the following error  message stating NFC reader connection the Stripe Terminal ConnectionToken has expired. Please download a new one using v1/terminal/connection_tokens

So I think I should call initTerminal on each payment, Unfortunately I couldn't able to call initTerminal multiple times , here I struck with the following error 
_You can only call initTerminal *before* requesting the Terminal instance for the first time. If you are trying to switch accounts in your app, refer to the documentation for the `clearCachedCredentials` method._

Without initTerminal, How can I generalte new token v1/terminal/connection_tokens
InitTerminal code I used is given below
    
    Terminal.initTerminal(
            applicationContext,
            LogLevel.VERBOSE,
            TokenProvider(tock), // token created from server for eg: stripe.terminal.connectionTokens.create();
            TerminalEventListener()
    )

Tokenprovider class is defined as given below

    class TokenProvider(
        var token: String
    ): ConnectionTokenProvider {
        override fun fetchConnectionToken(callback: ConnectionTokenCallback) {
            try { 
                callback.onSuccess(token)
            } catch (e: ConnectionTokenException) {
                callback.onFailure(e)
            }
        }
    }

After disconnectReader & clearCachedCredentials , I called Terminal.initTerminal as above, but receive error as below

_You can only call initTerminal *before* requesting the Terminal instance for the first time. If you are trying to switch accounts in your app, refer to the documentation for the `clearCachedCredentials` method._

**Comment by billfinn-stripe** [comment]
Can you please fill out the issue template and include any relevant stack traces?

**Comment by ajithv11** [comment]
I understood the issue and added a fix by adding dynamic token to fetchConnectionToken before I was using a static token while loading the app, SO for now I guess it is fixed, But some confusions are the could please have a look into my questions below

1. Terminal.initTerminal will call only once the app is loaded, It cannot call multiple times even After disconnectReader & clearCachedCredentials

2.  After disconnectReader & clearCachedCredentials after each success payments, the already initialised terminal will invoke fetchConnectionToken of ConnectionTokenProvider , So for each payment new token will generate. I hope this solve my issue.

Doubt : If point 2 I skip and if I use token throughout the system and if in a day there are few thousands of payments then can we use the same token , Does it have any expiry.. If expiry is there then my point 2 will solve it

Finally , If my point 2 is an ideal solution for tap to pay 

Thanks in advance

**Comment by ajithv11** [comment]
@billfinn-stripe  Thanks for your reply

One more question. What If I want to use same app to do multiple stripe connection for USD and GBP payment

Is it possible in this case ? Since I need to do Terminal.initTerminal with 2 different stripe accounts in same app
Is tap to pay designed for one android app for one stripe account(ie one currency support at a time)

In this case I worried about the error `You can only call initTerminal before requesting the Terminal instance for the first time.`

Please share your thoughts

**Comment by billfinn-stripe** [comment]
> In this case I worried about the error `You can only call initTerminal before requesting the Terminal instance for the first time.`

You should update your `fetchConnectionToken` to create a connection token for the correct account. When switching between accounts, be sure to call `clearCachedCredentials`. Switching accounts may also requires that you disconnect/reconnect to the reader. You should not call `initTerminal` multiple times.

**Comment by ajithv11** [comment]
@billfinn-stripe  That's cool :)

Now I got some idea about the workflow now, I ll try with your suggestion

Once again thanks a lot for explaining the workflow in detail

**Comment by billfinn-stripe** [comment]
I'm going to close this issue. If you encounter a bug or error, feel free to create a new issue after filling out the issue template.

Also, for questions, please start with emailing support@stripe.com, rather than filing Github issues. Github issues are meant to be for diagnosing stacktraces and filing bugs with the SDK, rather than answering questions about how one should integrate the SDK. Routing help requests through support@stripe.com helps us better identify and aggregate requests, so we can find missing or confusing pieces with our existing documentation.
